Re-factor props for `AnnotationBody` and add `AnnotationBody` to `AnnotationOmega`

lyzadanger opened the pull request:
This PR primarily moves some props around vis-a-vis `AnnotationBody` such that `Annotation` doesn't have to manage `Excerpt`'s state (its grandchild) via `AnnotationBody`. This allows for a simpler integration of `AnnotationBody` into `Annotation`.

This PR also makes a global change to `$grey-5`, making it a few ticks darker to be compliant with WCAG contrast-ratio requirements.

The new `More/Less` controls for expanding/collapsing excerpts in this PR are WCAG-compliant for contrast ratio.

Screen shot of new/more contrasty controls as of this PR:

![image](https://user-images.githubusercontent.com/439947/72738294-4566e000-3b6f-11ea-98ea-89892e12bf6a.png)


Part of #1650
Related to https://github.com/hypothesis/product-backlog/issues/1085

lyzadanger commented on src/sidebar/components/annotation-omega.js:
```diff
@@ -25,6 +26,18 @@ function AnnotationOmega({
   // TODO: `isEditing` will also take into account `isSaving`
   const isEditing = !!draft;
 
+  const annotationState = () =>
```
I may opt to create a store selector soon that returns draft-or-state for a given annotation; would consider this prototyped code for the moment.

lyzadanger commented:
The test coverage shortfall for `AnnotationOmega` is generally intentional. My next PR will focus on functionality in that component directly.

lyzadanger commented on src/sidebar/components/test/annotation-body-test.js:
```diff
@@ -30,4 +40,48 @@ describe('AnnotationBody', () => {
+      // change the `isCollapsible` state to `true` via the `Excerpt`
+      excerpt.props().onCollapsibleChanged(true);
+      // Force a re-render
+      wrapper.setProps({});
```
In this and the following test, I had trouble encouraging the element to re-render without using `setProps`. I'm not sure whether this is a code smell, an enzyme mis-use, or benign.

robertknight commented:
Some quick visual feedback:

- The new "More"/"Less" link looks better / is easier to see.
- I think there is too much space below the "More" text visually.
- The "More" button has a redundant `title` which exactly matches the text label. This causes an odd-looking tooltip to appear on desktop devices

<img width="436" alt="Screenshot 2020-01-20 17 04 52" src="https://user-images.githubusercontent.com/2458/72745185-13717000-3ba7-11ea-99f5-e2a24ef86e40.png">

lyzadanger commented on src/sidebar/components/annotation-body.js:
```diff
@@ -9,82 +13,78 @@ import MarkdownView from './markdown-view';
+
+  const onChangeCollapsed = toCollapsed => {
+    setIsCollapsed(toCollapsed);
+  };
```
I wasn't sure if it was kosher to pass around references to the state-setting functions that belong to the component—is it?

robertknight replied:
I'm not aware of any recommendations against it. The only theoretical issue I can see is that the behavior changes if the state setter is called with a function rather than a value ([see note](https://reactjs.org/docs/hooks-reference.html#functional-updates)). I don't see that being an issue here.

lyzadanger commented:
> This closure takes no arguments and isn't passed to anything else to call so you could just set annotationState to the value that the closure returns.

I'm going to take this up in my next body of work, which will be centered around `AnnotationOmega` annotation state.

lyzadanger commented on src/sidebar/components/test/annotation-body-test.js:
```diff
@@ -30,4 +40,48 @@ describe('AnnotationBody', () => {
+    act(() => {
+      // change the `isCollapsible` state to `true` via the `Excerpt`
+      excerpt.props().onCollapsibleChanged(true);
+      // Force a re-render
```
Thanks! All cleared up.

lyzadanger commented:
> I think there is too much space below the "More" text visually. I think the same amount above/below would work, but try it and see.

I agree. The remaining vertical space underneath the button is a function of `annotation-footer`'s `top-margin` of `15px`. I rather like the way the cards look without that margin, but I need more input before I make that change—can you try out turning off that `top-margin` and seeing what you think and whether there are layout cases I'm not thinking of where the tighter vertical rhythm could be problematic?

lyzadanger commented:
Thanks, @robertknight !

I've addressed your test and code feedback.

On the visual feedback: I've fixed the titles to be something sensible, carried over from the prior implementation. I'm going to leave the alignment thing alone for now. I have an open question to you about `top-margin` on `.annotation-footer`; that's the only open thing I know of here...

robertknight commented:
> I rather like the way the cards look without that margin, but I need more input before I make that change—can you try out turning off that top-margin and seeing what you think and whether there are layout cases I'm not thinking of where the tighter vertical rhythm could be problematic?

Yes - when you're editing an annotation:

<img width="443" alt="Screenshot 2020-01-21 08 33 18" src="https://user-images.githubusercontent.com/2458/72788047-ff6e5280-3c28-11ea-853a-0e793b26ae3c.png">

It also looks quite tight with a short annotation body:

<img width="448" alt="Screenshot 2020-01-21 08 33 52" src="https://user-images.githubusercontent.com/2458/72788099-1ca32100-3c29-11ea-9a4f-4bef13ad0280.png">

I tried changing the More/Less row 0.5em top padding (no bottom padding) and reducing the footer's top margin from 15 to 10px:

<img width="444" alt="Screenshot 2020-01-21 08 38 00" src="https://user-images.githubusercontent.com/2458/72788273-7f94b800-3c29-11ea-8eac-5516367ad8d3.png">

Compare this to the current state of this branch (at 2180dac9):

<img width="443" alt="Screenshot 2020-01-21 08 39 18" src="https://user-images.githubusercontent.com/2458/72788302-90ddc480-3c29-11ea-910e-30ea1090a4b4.png">

I experimented with a few variants like Jared's suggestion to overlay the More/Less "bar" on top of the bottom of the annotation body text with a top-to-bottom transparent-to-white gradient but it felt like this could make it hard to distinguish the "More/Less" link for some users.

lyzadanger commented:
@robertknight I've pushed an additional commit that I hope provides at least stop-gap vertical-rhythm improvements. Want to take another pass here before I merge?

robertknight commented:
I noticed the latest version of this branch adds some extra spacing when editing:

<img width="418" alt="lyza-adjustment" src="https://user-images.githubusercontent.com/2458/72822925-d2429400-3c6a-11ea-8dc0-0d5e9b95d5da.png">

Compare to the changes I mentioned in my previous comment, which kept everything structurally the same but reduced the annotation footer's top-margin from 15 to 10px and made the padding on the More/Less bar only top padding rather than top + bottom padding:

<img width="407" alt="Screenshot 2020-01-21 16 26 21" src="https://user-images.githubusercontent.com/2458/72822978-e5556400-3c6a-11ea-97c8-ab9539c68b69.png">

lyzadanger commented:
@robertknight The "extra space" when editing exists on master. The tighter vertical spacing in your second screenshot was introduced (and then backed out of) this PR.
